javacord-api/src/main/java/org/javacord/api/entity/message/Message.java

```
@@ -102,6 +104,41 @@ default MessageBuilder toMessageBuilder() {
         return api.getUncachedMessageUtil().delete(channelId, messageId, reason);
     }
 
+    /**
+     * Deletes the message after the given delay.
+     *
+     * @param time The time to be deleted after.
+     * @param unit The unit to measure the given {@code time} in.
+     * @return A future that completes once the message is deleted.
+     */
+    default CompletableFuture<Void> deleteAfter(long time, TimeUnit unit) {
+        return deleteAfter(time, unit, null);
+    }
+
+    /**
+     * Deletes the message after the given delay with the given reason.
+     *
+     * @param time The time to be deleted after.
+     * @param unit The unit to measure the given {@code time} in.
+     * @param reason The reason why the message has been deleted.
+     * @return A future that completes once the message is deleted.
+     */
+    default CompletableFuture<Void> deleteAfter(long time, TimeUnit unit, String reason) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        try {
+            getApi().getThreadPool()
+                    .getScheduler()
+                    .schedule(() -> this.delete(reason), time, unit)
+                    .get();
```

You shouldn't block the caller thread but properly use completable futures.

---

yes, i noticed that on  the last commit but i didnt have time to fix that yet

---

Fixed 👍

---

I wouldn't call that fixed, it is like if you call `invokeLater` just because you have strange behavior in Swing or like calling `Thread.sleep` just because a multi-threaded program behaves strange. Wrapping this stuff in yet another thread is no just a workaround, not a proper solution.

Besides that, you also ignore any exceptions happening in the `delete`.

---

I cannot properly handle checked exceptions in SupplyAsync without rethrowing them underlying a runtime exception..... What do you propose I do

---

There is no need for checked exceptions?

---

And I don't see any supply sync

---

How else do I wait for the ScheduledFuture to be completed after deletion than using #get? I could return the ScheduledFuture instead of the CompletableFuture, that might be even better.

---

The better question is, why should you wait? It is an asynchronous process.

---

Because `ScheduledFuture` offers no callbacks and the least disgusting way (that I'm aware of) of getting from a `ScheduledFuture` to a `CompletableFuture` is wrapping the former in the latter via `CompletableFuture#supplyAsync` or similar means.

---

I'm going with just returning the `ScheduledFuture` as that is the most logical thing to do with a scheduled task.